compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/TextFieldScroll.uikit.kt

```diff
@@ -18,9 +18,182 @@ package androidx.compose.foundation.text
+import androidx.compose.ui.text.input.TransformedText
+import androidx.compose.ui.text.input.VisualTransformation
+import androidx.compose.ui.unit.Constraints
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.unit.LayoutDirection
+import kotlin.math.min
+import kotlin.math.roundToInt
 
 @ExperimentalFoundationApi
 @Composable
-internal actual fun rememberTextFieldOverscrollEffect(): OverscrollEffect? = rememberOverscrollEffect()
\ No newline at end of file
+internal actual fun rememberTextFieldOverscrollEffect(): OverscrollEffect? =
+    rememberOverscrollEffect()
+
+internal actual fun Modifier.textFieldScroll(
+    scrollerPosition: TextFieldScrollerPosition,
+    textFieldValue: TextFieldValue,
+    visualTransformation: VisualTransformation,
+    textLayoutResultProvider: () -> TextLayoutResultProxy?
+): Modifier {
+    val orientation = scrollerPosition.orientation
+    val cursorOffset = scrollerPosition.getOffsetToFollow(textFieldValue.selection)
+    scrollerPosition.previousSelection = textFieldValue.selection
+
+    val transformedText = visualTransformation.filterWithValidation(textFieldValue.annotatedString)
+
+    val layout = when (orientation) {
+        Orientation.Vertical ->
+            IOSVerticalScrollLayoutModifier(
+                scrollerPosition,
+                cursorOffset,
+                transformedText,
+                textLayoutResultProvider
+            )
+
+        Orientation.Horizontal ->
+            IOSHorizontalScrollLayoutModifier(
+                scrollerPosition,
+                cursorOffset,
+                transformedText,
+                textLayoutResultProvider
+            )
+    }
+    return this.clipToBounds().then(layout)
+}
+
+private data class IOSVerticalScrollLayoutModifier(
+    val scrollerPosition: TextFieldScrollerPosition,
+    val cursorOffset: Int,
+    val transformedText: TransformedText,
+    val textLayoutResultProvider: () -> TextLayoutResultProxy?
+) : LayoutModifier {
+    override fun MeasureScope.measure(
+        measurable: Measurable,
+        constraints: Constraints
+    ): MeasureResult {
+        val childConstraints = constraints.copy(maxHeight = Constraints.Infinity)
+        val placeable = measurable.measure(childConstraints)
+        val height = min(placeable.height, constraints.maxHeight)
+
+        return layout(placeable.width, height) {
+            val textLayoutResult = textLayoutResultProvider()?.value
+            val cursorRect = getCursorRectInScroller(
+                cursorOffset = cursorOffset,
+                transformedText = transformedText,
+                textLayoutResult = textLayoutResult,
+                rtl = false,
+                textFieldWidth = placeable.width
+            )
+
+            val additionalDeltaY = if (textLayoutResult != null) {
+                // On iOS we will scroll to one line before the cursor:
+                val scrollToLine = textLayoutResult.getLineForOffset(cursorOffset) - 1
```

Comment:
Have you checked it with a case when the cursor is in the first position (line 0)?

Reply:
Yes. It checks with coerceIn() function in next lines

Reply:
But I can move this check to current line.

Reply:
Refactored thic block of code for better understanding